src/swtpm/swtpm_nvstore_linear.h

```diff
@@ -0,0 +1,107 @@
+    uint32_t      length;
+    unsigned char *data;
+
+    struct nvram_linear_hdr *hdr; /* points into *data */
+};
+
+/*
+    Implementation ops for linear data backends. The "linear" backend takes care
+    of the allocation strategy to pack multiple files/parts into one address
+    space, implementations only need to make sure that the linearized data is
+    written and accessed correctly.
+*/
+struct nvram_linear_store_ops {
+    /*
+        Called once upon initialization, if the URI prefix (e.g. file://)
+        matches the implementation. 'uri' is the raw, prefixed backend_uri
+        string. 'data' and 'length' should be set to a memory region that
+        contains the loaded data in it's entirety (e.g. when loaded from a file,
+        the mmap base address and file length).
+        If a new store is created, it must contain at least enough space to
+        store 'sizeof(struct nvram_linear_hdr)' bytes.
+    */
+    TPM_RESULT (*open)(const char* uri,
+                       unsigned char **data,
+                       uint32_t *length);
+
+    /*
+        Called whenever the data in the provided buffer has changed. Will be
+        called for every changed region, offset is relative to base (0 is for
+        header or full flushes). Can be left unimplemented if data stored in
+        given buffer is flushed automatically.
+    */
+    TPM_RESULT (*flush)(const char* uri,
+                        uint32_t offset,
+                        uint32_t count);
+
+    /*
+       Called whenever space has been freed or more space is required to store
+       data. Implementations can choose to leave this unimplemented, or make the
+       implementation a no-op, TPM_SIZE should be returned if 'requested_length'
+       can not be made available. 'data' and 'new_length' must be set similar to
+       open(), either to the same region or a different one (in case of remap or
+       similar).
+    */
+    TPM_RESULT (*resize)(const char* uri,
+                         unsigned char **data,
+                         uint32_t *new_length,
+                         uint32_t requested_length);
+};
+
+TPM_RESULT
+SWTPM_NVRAM_Prepare_Linear(const char *uri);
+
+TPM_RESULT
+SWTPM_NVRAM_LoadData_Linear(unsigned char **data,
+                            uint32_t *length,
+                            uint32_t tpm_number,
+                            const char *name,
+                            const char *uri);
+
+TPM_RESULT
+SWTPM_NVRAM_StoreData_Linear(unsigned char *filedata,
+                             uint32_t filedata_length,
+                             uint32_t tpm_number,
+                             const char *name,
+                             const char *uri);
+
+TPM_RESULT
+SWTPM_NVRAM_DeleteName_Linear(uint32_t tpm_number,
+                              const char *name,
+                              TPM_BOOL mustExist,
+                              const char *uri);
+
+extern struct nvram_backend_ops nvram_linear_ops;
```

Comment:
I suppose most of the structures are private can be moved into the .c file except for this last line here that we may sooner or later have to consolidated into a common .h file. See my cleanup patch in PR #556.